content/posts/i-like-rust.md

```diff
@@ -0,0 +1,218 @@
+Compile times have improved over the last years and they will continue to improve.
+If you have issues with compilation times there are [many ways to further improve them](https://corrode.dev/blog/tips-for-faster-rust-compile-times/).
+
+## Hard to learn
+
+The Rust compiler is strict, that's no secret. As a beginner it can be difficult to get basic programs to compile. Understanding the concepts of the language such as ownership is therefore essential. Fortunately, the official [Rust book](https://doc.rust-lang.org/stable/book/) is perfect for learning all these concepts, and the Rust compiler will also help you along the way.
+
+You will encounter compilation errors probably more often than in any other language.
+There will be cases where you have no idea how to fix the problem at first.
+Eventually, you might realise that you will have to rethink your whole approach to the problem.
+
+This was the hardest part for me (and still is) when working with Rust. Sometimes the compiler forces you to rethink your approach. You realise that you can't bend the borrow checker to your will, instead you have to work with it. The compiler is your mentor not your enemy to fight.
+
+Personally, I started learning Rust during the Covid pandemic. At the time we had a course on algorithms and data structures at university and I decided to implement the exercises in Rust instead of Java. After reading the first few chapters of the Rust book, I felt confident enough to give it a try. I was able to implement some algorithms and was happy with the results. But I got lost and confused when I tried to implement a binary tree myself. After trail and error, I did some research and came across a chapter in the book I hadn't read yet. Turns out that data structures in Rust can get complicated very quickly. On top of ownership you also need to know about smart pointers, interior mutability, reference counting, reference cycles...
+I was overwhelmed and discouraged. But I realised that data structures were not the best starting point for learning Rust. And that in the real world you would make use of the standard library and crates built by smart people anyway.
+
+After that, I tried to implement simple things with Rust. Over time I almost forgot about the concepts I read about when dealing with binary trees. It turns out that programming "normal" things in Rust is much easier than I though. The only thing you really need to understand is ownership. Understanding lifetimes, interior mutability, etc. is great, but in "normal" code you almost never make use of these concepts.
+
+## Conclusion
+
+In my opinion, Rust is the pinnacle of modern compiler and language development.
+The deliberate design choices in the language show how the developers have learned from the problems in other languages.
+The introduction of new concepts such as the borrow checker make the Rust compiler probably the most complex compiler out there.
```

Comment:
Hm, maybe? Then again, the C++ compiler is probably also pretty complex. Also, `rustc` is just a compiler frontend for LLVM, so that would be worth a mention as well. ¯\_(ツ)_/¯